roon-server: disable DynamicUser

reanimus opened the pull request:
###### Motivation for this change
DynamicUser currently breaks the backup functionality provided by roon,
as the roon server cannot write to non-canonical directories and the
recycled UIDs/GIDs would make managing permissions for the directory
impossible. On top of that, it would break the ability to manage the
local music library files (as it would not be able to delete them).

###### Things done


- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

reanimus committed:
roon-server: disable DynamicUser

DynamicUser currently breaks the backup functionality provided by roon,
as the roon server cannot write to non-canonical directories and the
recycled UIDs/GIDs would make managing permissions for the directory
impossible. On top of that, it would break the ability to manage the
local music library files (as it would not be able to delete them).

Infinisil commented:
You should use a normal NixOS user declaration instead then. How it is now the service would run as `root` which is no good

reanimus commented:
Yeah I figured I need to do some sort of id declaration. adding it in now

reanimus committed:
roon-server: Add actual user piping

Adds defined IDs

Infinisil commented on nixos/modules/misc/ids.nix:
```diff
@@ -340,6 +340,7 @@
       cockroachdb = 313;
       zoneminder = 314;
       paperless = 315;
+      roon-server = 316;
```
Ah, no need for that though. NixOS will automatically assign an {u,g}id if you don't set one.

Infinisil replied:
(this is what https://github.com/NixOS/nixpkgs/pull/65698 is aiming for)

reanimus replied:
I'm curious: what scenario would specifically require a fixed predefined UID/GID for everyone if nix can just assign one? o:

Infinisil replied:
I can only think of some kind of service that needs to interact with other instances on different machines by means of moving files between them, and for some reason not having implemented uid adjustments. IIRC an old version of NFS might have been such a case.

reanimus committed:
roon-server: let nix assign ids

reanimus commented:
Also, should I set `isSystemUser = true;` for the roon server user?

Infinisil commented:
Yeah I'd set `isSystemUser = true`

reanimus committed:
[roon-server] make roon user a system user

aanderse commented on nixos/modules/services/audio/roon-server.nix:
```diff
@@ -33,9 +47,7 @@ in {
-        DynamicUser = true;
-        SupplementaryGroups = "audio";
-        StateDirectory = name;
+        User = cfg.user;
```
Don't forget `Group = cfg.group;`!

Infinisil replied:
Actually, I believe systemd sets the group to be the primary group of the user by default. But I also prefer having it set here

aanderse reviewed (changes requested):
Is there an explicit reason to make the `user` and `group` configurable instead of hard coding?

reanimus committed:
[roon-server] don't create user if user changes defaults

If the user changes the user for roon, we can assume they handled the setup for it

reanimus commented:
There's not a specific reason for customizing it, no. It can be hard-coded.

reanimus commented:
Actually, I think it would be useful at make managing permissions easier, I think. Roon itself is usually packaged to run as root, which means it usually ignores a lot of this (and it makes some assumptions about things it can do)

Infinisil commented on nixos/modules/services/audio/roon-server.nix:
```diff
@@ -19,6 +19,20 @@ in {
         '';
       };
+      user = mkOption {
+        type = types.string;
```
Use `types.str` instead. `types.string` is deprecated

reanimus committed:
[roon-server] Use non-deprecated string type

Infinisil commented on nixos/modules/services/audio/roon-server.nix:
```diff
@@ -33,9 +47,8 @@ in {
         LimitNOFILE = 8192;
-        DynamicUser = true;
-        SupplementaryGroups = "audio";
-        StateDirectory = name;
```
Just noticed this, why remove this?

reanimus replied:
It was complaining on startup when I left the StateDirectory bit in there. Not sure why.

Infinisil replied:
Without this, there can't really be any place for it to store data, so this should definitely be looked into further

reanimus replied:
I removed it initially because it threw an error with it when I tested before. After some more troubleshooting, I think it's because I left the directory structure the same as when it was dynamic user based. In this configuration, the actual directory lives under `/var/lib/private` and the path `/var/lib/roon-server` is a symlink to the `roon-server` subdirectory in private. Moving it into the actual place fixed the error on launch.

Infinisil replied:
Hm I see, that makes me worried that systemd doesn't handle `DynamicUser` -> normal user transitions at all..

I guess it's fine anyways for this service since it's so new and didn't reach stable yet

reanimus committed:
roon-server: add back state directory

reanimus commented:
Anything else I ought to clean up here?

Infinisil commented:
Yeah looks good
